Bootstrap

stan-osipov opened the pull request:
When `com.onesignal.unity.bootstrap` is added to the project it will:

1. Add following dependencies to the project `manifest.json`
    * `com.onesignal.unity.ios`
    * `com.onesignal.unity.android`
2. Add NPMJS OneSgianl Package Registry
```
"name": "npmjs",
"url": "https://registry.npmjs.org/",
"scopes": [
     "com.onesignal"
 ]
```
3. Remove itself from the project
    * We want to keep user project as clean as possible
    * We want users to be able to delete any of the installed products (bootstrapper will interfere if exists)
   *  Bootstrapper does not contain any public C# API and useless for the user.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/onesignal/onesignal-unity-sdk-bootstrap/3)
<!-- Reviewable:end -->

alexey-yaremenko committed:
bootstrapper implementation

stan-osipov committed:
meta udpated

stan-osipov committed:
Bootstrapper self removal added.

stan-osipov committed:
minor cleanup

stan-osipov committed:
Namespace & scope registry name updated.

kostiantyn-koretskyi committed:
Update package.json

stan-osipov committed:
publish-to-npm

stan-osipov committed:
Merge branch 'master' of github.com:StansAssets/OneSignal-Unity-SDK-Bootstrap into master

kostiantyn-koretskyi committed:
version updated to 0.0.1-preview

stan-osipov committed:
chore: minor fixes and version bump

kostiantyn-koretskyi committed:
Update package.json

yourneko committed:
feat: the Bootstrapper downloads the latest release version of packages instead of a predefined one
feat: lock file that prevents an execution of the Bootstrapper

kostiantyn-koretskyi committed:
Merge pull request #1 from yourneko/repository-latest-release-version

Getting latest release version from github repository

stan-osipov committed:
chore: Installation fine tune

yourneko committed:
chore: '-test' postfixes in package names removed

yourneko committed:
feat: Automatic removal of directories of outdated SDK versions. Confirmation dialogue that allows user to interrupt the installation process

stan-osipov committed:
Update package.json

jkasten2 commented on CHANGELOG.md:
```diff
@@ -0,0 +1,12 @@
+# Change Log
```
Omit `CHANGELOG.md`, we will store these as github release notes instead.

jkasten2 commented on Editor/BootstrapperTestWindow.cs:
```diff
@@ -0,0 +1,30 @@
+        [MenuItem("OneSignalBootstrap/TestWindow")]
+        static void Init ()
+        {
+            var window = (BootstrapperTestWindow)EditorWindow.GetWindow (typeof (BootstrapperTestWindow));
```
nit formating, remove space before `(` on this line, and others where this issue exists

jkasten2 commented on Editor/BootstrapperTestWindow.cs:
```diff
@@ -0,0 +1,30 @@
+    {
+        static Label s_LatestVersionLabel;
+
+        [MenuItem("OneSignalBootstrap/TestWindow")]
```
Recommend renaming to "DebugWindow" if we plan to keep this.

jkasten2 commented on Editor/Core/GitHubRelease.cs:
```diff
@@ -0,0 +1,31 @@
+        internal string Name { get; }
+        internal string Tag { get; }
+
+        internal GitHubRelease (string json)
```
nit formating, remove space before `(` on this line, and others where this issue exists

jkasten2 commented on Editor/Core/GitHubRelease.cs:
```diff
@@ -0,0 +1,31 @@
+            Tag = FindValueByKey ("tag_name", json);
+        }
+
+        string FindValueByKey (string key, string json)
```
Could we add a comment of the expected format of key and json? Also an example of how the method should be called.

jkasten2 replied:
The implementation of this method looks very fragile. Could we use a JSON parser (such as Json.Deserialize) instead of raw string operations?

jkasten2 commented on Editor/Core/Manifest.cs:
```diff
@@ -0,0 +1,166 @@
+        {
+            Path = pathToFile;
+            m_ScopeRegistries = new Dictionary<string, ScopeRegistry>();
+            m_Dependencies = new Dictionary<string,Dependency>();
```
nit, missing space after `,`

jkasten2 commented on Editor/BootstrapperTestWindow.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace Com.OneSignal.Bootstrapper
+{
+    public class BootstrapperTestWindow : EditorWindow
```
Normally the bootstrapper gets removed right away, when would this be used? It seems like this window might be helpful in debugging any issues someone might have installing OneSignal if the bootstrapper doesn't work. Would be helpful to add a comment here on what things this test / debug window could help a with when there is an issue.

jkasten2 commented on Editor/Core/Manifest.cs:
```diff
@@ -0,0 +1,166 @@
+            m_RawContent[k_DependenciesKey] = dependencies;
+
+            string manifestText = Json.Serialize(m_RawContent, true);
+            File.WriteAllText(Path,manifestText);
```
nit formating, missing space after `,`

jkasten2 commented on Editor/Core/Manifest.cs:
```diff
@@ -0,0 +1,166 @@
+        /// </summary>
+        /// <param name="url">ScopeRegistry url to search for.</param>
+        /// <returns>`true` if scoped registry found, `false` otherwise.</returns>
+        public bool IsRegistryExists(string url)
```
`IsRegistryExists` isn't grammatically correct. Recommended `isRegistryPresent`

jkasten2 commented on Editor/Core/Manifest.cs:
```diff
@@ -0,0 +1,166 @@
+        /// </summary>
+        /// <param name="name">The dependency name to search for.</param>
+        /// <returns>`true` if found, `false` otherwise.</returns>
+        public bool IsDependencyExists(string name)
```
`IsDependencyExists` isn't grammatically correct. Recommended `IsDependencyPresent`

jkasten2 commented on Editor/Core/ScopeRegistry.cs:
```diff
@@ -0,0 +1,121 @@
+        /// <returns>Hash of this object.</returns>
+        public override int GetHashCode() {
+            int hash = 0;
+            if (!string.IsNullOrEmpty(Url)) hash ^= Url.GetHashCode();
```
This line is hard to read. Recommend splitting it up.
```C#
if (!string.IsNullOrEmpty(Url))
    hash ^= Url.GetHashCode();
```

jkasten2 commented on Editor/Core/ScopeRegistry.cs:
```diff
@@ -0,0 +1,121 @@
+        public override int GetHashCode() {
+            int hash = 0;
+            if (!string.IsNullOrEmpty(Url)) hash ^= Url.GetHashCode();
+             if (Scopes != null) {
```
Indentation is off on this if block

jkasten2 commented on Editor/Utilities/CleanUpUtility.cs:
```diff
@@ -0,0 +1,27 @@
+using UnityEditor;
```
Missing:
*  `using System.Collections.Generic;`
* `using UnityEngine;`

jkasten2 commented on Editor/Utilities/CleanUpUtility.cs:
```diff
@@ -0,0 +1,27 @@
+                    failedPathsList.ForEach(Debug.Log);
+                }
+                else {
+                    Debug.Log ($"Folder {RemovalPath} not found!");
```
It doesn't seem `RemovalPath` is defined.

yourneko committed:
chore: removed spacing before parentheses

yourneko committed:
fix: Updated code and usings for UNITY_2020.

chore: Code formatting aligned with OneSignal's standards;
Removed CHANGELOG.md due to OneSignal's request;
GitHubUtility returns string value instead of GitHubRelease. Home-brewed FindValueByKey method removed, json is parsed with MiniJSON utility instead;
Minor changes in Bootstrapper.cs to improve the readability;
'lock' file renamed to 'onesignal-bootstrap-lock'. It should be placed in Assets/ of the test project to make a .unitypackage

stan-osipov committed:
Merge pull request #2 from yourneko/master

Removed unneeded EditorWindow, fixed formatting and naming, fixed usings and issues for UNITY_2020

yourneko committed:
chore: removed 'using' directive

jkasten2 reviewed (changes requested):
### Update
Confirmed this resolved.

"Scope(s)" should be "com.onesignal.unity" not "com.onesignal" this way when OneSignal adds non-unity packages to NPMJS in the future it won't show up in Unity.

![image](https://user-images.githubusercontent.com/645861/112355594-0bb9c980-8c8b-11eb-9b61-66e49538e79b.png)

jkasten2 reviewed (changes requested):
### Update
Once we have a release on Github with a matching published NPMJS version this should not be happening..

Getting this error after closing the Package Manager window after copying OneSignal-Unity-SDK-Bootstrap into the project.
```
An error occurred while resolving packages:
  Project has invalid dependencies:
    com.onesignal-test.unity.android: Version 'xx' is invalid. Expected one of: a 'SemVer' compatible value; a value starting with 'file:'; a Git URL starting with 'git:' or 'git+', or ending with '.git'.
    com.onesignal-test.unity.ios: Version 'xx' is invalid. Expected one of: a 'SemVer' compatible value; a value starting with 'file:'; a Git URL starting with 'git:' or 'git+', or ending with '.git'.

A re-import of the project may be required to fix the issue or a manual modification of D:/temp/OneSignal-Unity-SDK-Bootstrap-test/Packages/manifest.json file.
```


Relevant parts of `/manifest.json`:
```{
  "dependencies": {
    "com.onesignal-test.unity.android": "xx",
    "com.onesignal-test.unity.ios": "xx"
  },
  "scopedRegistries": [
    {
      "name": "npmjs",
      "url": "https://registry.npmjs.org/",
      "scopes": [
        "com.onesignal-test"
      ]
    }
  ]
}
```

It isn't clear where it is pulling `xx` from. I tested with `com.onesignal-test` and `com.onesignal` scopes.

jkasten2 reviewed (changes requested):
TODO:
Retest this one before release, it depends on com.onesignal.unity.core being available.

It seems the Bootsrap isn't removing itself. I could be due to the other error I was getting noted above but even after I fixed the versions in the `manifest.json` it still won't clean it self up.

![image](https://user-images.githubusercontent.com/645861/112358644-c4810800-8c8d-11eb-8ede-dad1c182630b.png)

I ensured the folder was placed in the correct spot per the code:
![image](https://user-images.githubusercontent.com/645861/112362073-49215580-8c91-11eb-8974-dce0b10a97be.png)

I tested on Unity 2020.3.1f1 LTS on Windows 10 if this matters

jkasten2 commented on Documentation~/com.onesignal.unity.bootstrap.md:
```diff
@@ -0,0 +1 @@
+Full documentation avaliable at  [documentation.onesignal.com](https://documentation.onesignal.com/docs/unity-sdk-setup)
```
Spelling, "avaliable" should be "available"

jkasten2 commented on Editor/BootstrapperConfig.cs:
```diff
@@ -16,6 +16,11 @@ static class BootstrapperConfig
 
         public const string GitHubRepositoryURL = @"ssh://git@github.com:StansAssets/OneSignal-Unity-SDK.git";
         public const string BootstrapperFolderPath = @"Assets/OneSignalBootstrap/Editor";
```
Shouldn't this clean up the whole `Assets/OneSignalBootstrap` folder, not just the `Editor`?

stan-osipov replied:
agreed

jkasten2 commented on Editor/BootstrapperConfig.cs:
```diff
@@ -16,6 +16,11 @@ static class BootstrapperConfig
 
         public const string GitHubRepositoryURL = @"ssh://git@github.com:StansAssets/OneSignal-Unity-SDK.git";
```
Just noting here so we don't forget. Replace `StansAssets` with `OneSignal` here before releasing.

stan-osipov replied:
This has to be done in this PR actually.  Thx for heads up.

stan-osipov committed:
chore: PR feedback midnor fixed.

stan-osipov committed:
chore: config fix
